se_denseslam/include/se/rendering.hpp

```diff
@@ -0,0 +1,202 @@
+        if (surfNorm.norm() == 0) {
+          //normal[pos] = normalize(surfNorm); // APN added
+          normal[pos.x() + pos.y() * normal.width()] = Eigen::Vector3f(INVALID, 0, 0);
+        } else {
+          // Invert normals if SDF
+          normal[pos.x() + pos.y() * normal.width()] = std::is_same<T, SDF>::value ?
+            (-1.f * surfNorm).normalized() : surfNorm.normalized();
+        }
+      } else {
+        vertex[pos.x() + pos.y() * vertex.width()] = Eigen::Vector3f::Constant(0);
+        normal[pos.x() + pos.y() * normal.width()] = Eigen::Vector3f(INVALID, 0, 0);
+      }
+    }
+  TOCK("raycastKernel", vertex.width() * vertex.height());
+}
+
+// void renderNormalKernel(uchar3* out, const float3* normal, uint2 normalSize) {
+// 	TICK();
+// 	unsigned int y;
+// #pragma omp parallel for shared(out), private(y)
+// 	for (y = 0; y < normalSize.y; y++)
+// 		for (unsigned int x = 0; x < normalSize.x; x++) {
+// 			unsigned int pos = (x + y * normalSize.x);
+// 			float3 n = normal[pos];
+// 			if (n.x == -2) {
+// 				out[pos] = make_uchar3(0, 0, 0);
+// 			} else {
+// 				n = normalize(n);
+// 				out[pos] = make_uchar3(n.x * 128 + 128, n.y * 128 + 128,
+// 						n.z * 128 + 128);
+// 			}
+// 		}
+// 	TOCK("renderNormalKernel", normalSize.x * normalSize.y);
+// }
+
+void renderDepthKernel(unsigned char*         out,
+                       float*                 depth,
+                       const Eigen::Vector2i& depthSize,
+                       const float            nearPlane,
+                       const float            farPlane);
+
+void renderTrackKernel(unsigned char*         out,
+                       const TrackData*       data,
+                       const Eigen::Vector2i& outSize);
+
+template <typename T>
+void renderVolumeKernel(const Volume<T>&                  volume,
+                        unsigned char*                    out, // RGBW packed
+                        const Eigen::Vector2i&            depthSize,
+                        const Eigen::Matrix4f&            view,
+                        const float                       nearPlane,
+                        const float                       farPlane,
+                        const float                       mu,
+                        const float                       step,
+                        const float                       largestep,
+                        const Eigen::Vector3f&            light,
+                        const Eigen::Vector3f&            ambient,
+                        bool                              render,
+                        const se::Image<Eigen::Vector3f>& vertex,
+                        const se::Image<Eigen::Vector3f>& normal) {
+  TICK();
+#pragma omp parallel for
+  for (int y = 0; y < depthSize.y(); y++) {
+    for (int x = 0; x < depthSize.x(); x++) {
+      Eigen::Vector4f hit;
+      Eigen::Vector3f test, surfNorm;
+      const int idx = (x + depthSize.x()*y) * 4;
+
+      if (render) {
```

Comment:
Since we are cleaning this up... I am not sure I like this if here, but why don't we call this variable something like `raycast_normals` instead of `render`?

Reply:
Renamed to raycast_normals and also renamed a few other variables.